crates/madara/primitives/gateway/src/user_transaction.rs

```diff
@@ -38,11 +40,52 @@ use serde::{Deserialize, Serialize};
 use serde_with::serde_as;
 use starknet_types_core::felt::Felt;
 use starknet_types_rpc::{
-    BroadcastedDeclareTxn, BroadcastedDeclareTxnV1, BroadcastedDeclareTxnV2, BroadcastedDeclareTxnV3,
-    BroadcastedDeployAccountTxn, BroadcastedInvokeTxn, BroadcastedTxn, DeployAccountTxnV1, DeployAccountTxnV3,
-    InvokeTxnV0, InvokeTxnV1, InvokeTxnV3,
+    AddInvokeTransactionResult, BroadcastedDeclareTxn, BroadcastedDeclareTxnV1, BroadcastedDeclareTxnV2,
+    BroadcastedDeclareTxnV3, BroadcastedDeployAccountTxn, BroadcastedInvokeTxn, BroadcastedTxn,
+    ClassAndTxnHash as AddDeclareTransactionResult, ContractAndTxnHash as AddDeployAccountTransactionResult,
+    DeployAccountTxnV1, DeployAccountTxnV3, InvokeTxnV0, InvokeTxnV1, InvokeTxnV3,
 };
 
+/// Gateway response when a transaction is successfully added to the mempool.
+/// Generic type T represents the specific transaction result type
+/// (Invoke, Declare, or DeployAccount).
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct AddTransactionResult<T> {
+    /// Status code indicating the transaction was received successfully
+    code: AddTransactionCode,
+    /// The transaction result
+    #[serde(flatten)]
+    result: T,
+}
+
+/// Marker trait to restrict which types can be used as transaction results.
+/// Only valid Starknet transaction responses implement this trait.
+pub trait ValidTransactionResult {}
+
+// Implement the marker trait for the three valid transaction result types
+impl<T> ValidTransactionResult for AddInvokeTransactionResult<T> {}
+impl<T> ValidTransactionResult for AddDeclareTransactionResult<T> {}
+impl<T> ValidTransactionResult for AddDeployAccountTransactionResult<T> {}
```

Comment:
Consider sealing this trait by either making it private if does not need to be accessed outside of this crate, or a sub-trait of a private trait if it does.

```suggestion
/// Ensures [ValidTransactionResult] is accessible from outside the crate but cannot be implemented on any new structs
trait Sealed {} 

/// Marker trait to restrict which types can be used as transaction results.
/// Only valid Starknet transaction responses implement this trait.
pub trait ValidTransactionResult: Sealed {}

// Implement the marker trait for the three valid transaction result types
impl<T> Sealed for AddInvokeTransactionResult<T> {}
impl<T> Sealed for AddDeclareTransactionResult<T> {}
impl<T> Sealed for AddDeployAccountTransactionResult<T> {}
impl<T> ValidTransactionResult for AddInvokeTransactionResult<T> {}
impl<T> ValidTransactionResult for AddDeclareTransactionResult<T> {}
impl<T> ValidTransactionResult for AddDeployAccountTransactionResult<T> {}
```